File: eval_harness/utils/shell/timeout_exec.py
#!/usr/bin/env python3
"""Run a command with a timeout, including on macOS where GNU timeout is absent."""

from __future__ import annotations

import os
import signal
import subprocess
import sys


def main() -> int:
    if len(sys.argv) < 3:
        print("usage: timeout_exec.py <seconds> <cmd> [args...]", file=sys.stderr)
        return 2
    try:
        seconds = float(sys.argv[1])
    except ValueError:
        print(f"invalid timeout: {sys.argv[1]}", file=sys.stderr)
        return 2

    proc = subprocess.Popen(sys.argv[2:], start_new_session=True)
    try:
        return proc.wait(timeout=seconds)
    except subprocess.TimeoutExpired:
        print(
            f"timeout_exec.py: command exceeded {seconds:g}s; terminating process group",
            file=sys.stderr,
        )
        try:
            os.killpg(proc.pid, signal.SIGTERM)
            proc.wait(timeout=10)
        except Exception:
            try:
                os.killpg(proc.pid, signal.SIGKILL)
            except ProcessLookupError:
                pass
        return 124


if __name__ == "__main__":
    raise SystemExit(main())

File: eval_harness/utils/shell/timeout_exec.ts
#!/usr/bin/env bun

import { constants } from "node:os";

const USAGE = "usage: timeout_exec.ts <seconds> <cmd> [args...]";
const MAX_TIMER_MILLISECONDS = 2_147_483_647;
const TIMEOUT_PATTERN = /^\+?(?:[0-9]+(?:\.[0-9]*)?|\.[0-9]+)(?:[eE][+-]?[0-9]+)?$/;

type ExitOrTimeout =
  | { kind: "exit"; exitCode: number }
  | { kind: "timeout" };

function parseTimeout(text: string): { seconds: number; milliseconds: number } | undefined {
  const stripped = text.trim();
  if (!TIMEOUT_PATTERN.test(stripped)) return undefined;
  const seconds = Number(stripped);
  const milliseconds = seconds * 1_000;
  if (!Number.isFinite(milliseconds) || milliseconds > MAX_TIMER_MILLISECONDS) {
    return undefined;
  }
  return { seconds, milliseconds };
}

async function waitForExitOrTimeout(
  exited: Promise<number>,
  timeoutMilliseconds: number,
): Promise<ExitOrTimeout> {
  let timeoutId: ReturnType<typeof setTimeout> | undefined;
  const timeout = new Promise<ExitOrTimeout>((resolve) => {
    timeoutId = setTimeout(() => resolve({ kind: "timeout" }), timeoutMilliseconds);
  });

  const outcome = await Promise.race([
    exited.then((exitCode) => ({ kind: "exit", exitCode }) as const),
    timeout,
  ]);

  if (timeoutId !== undefined) {
    clearTimeout(timeoutId);
  }
  return outcome;
}

function forceKillProcessGroup(pid: number): void {
  try {
    process.kill(-pid, "SIGKILL");
  } catch {
    // Best-effort: Bun may already have reaped the group, and macOS can report
    // either ESRCH or EPERM for this race. Cleanup must not replace exit 124.
  }
}

export async function main(args: string[] = process.argv.slice(2)): Promise<number> {
  if (args.length < 2) {
    console.error(USAGE);
    return 2;
  }

  const timeoutText = args[0];
  if (timeoutText === undefined) {
    console.error(USAGE);
    return 2;
  }

  const timeout = parseTimeout(timeoutText);
  if (timeout === undefined) {
    console.error(`invalid timeout: ${timeoutText}`);
    return 2;
  }

  const command = args.slice(1);
  const child = Bun.spawn(command, {
    stdin: "inherit",
    stdout: "inherit",
    stderr: "inherit",
    detached: true,
  });

  const outcome = await waitForExitOrTimeout(child.exited, timeout.milliseconds);
  if (outcome.kind === "exit") {
    if (child.signalCode !== null) {
      return -constants.signals[child.signalCode];
    }
    return outcome.exitCode;
  }

  console.error(
    `timeout_exec.ts: command exceeded ${timeout.seconds}s; terminating process group`,
  );

  let forceKill = false;
  try {
    process.kill(-child.pid, "SIGTERM");
    const termination = await waitForExitOrTimeout(child.exited, 10_000);
    forceKill = termination.kind === "timeout";
  } catch {
    forceKill = true;
  }

  if (forceKill) {
    forceKillProcessGroup(child.pid);
  }
  return 124;
}

if (import.meta.main) {
  process.exit(await main());
}
